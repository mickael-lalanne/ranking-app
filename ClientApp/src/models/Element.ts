export interface Element {
    id?: string;
    name: string;
    image: string;
    type?: number;
}

export interface RankedElement {
    elementId: string;
    tierId: string;
    position: number;
}
