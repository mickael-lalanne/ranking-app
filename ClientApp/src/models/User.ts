export type UserId = string | null | undefined;

export interface User {
    id: UserId;
}
