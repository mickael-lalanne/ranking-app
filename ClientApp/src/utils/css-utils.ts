export const ELEMENT_SIZE: string = '75px';
