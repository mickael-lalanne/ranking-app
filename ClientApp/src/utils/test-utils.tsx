import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';
import type { AppStore, RootState } from '../app/store';
import { ThemeOptions, ThemeProvider, createTheme } from '@mui/material/styles';
import '@testing-library/jest-dom';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

// Mock MUI theme
const RANKING_APP_THEME_MOCK: ThemeOptions = {
    defaultRankingTheme: {
        primary: '',
        primaryLight: '',
        light: '',
        dark: '',
        info: '',
    },
};
const MockTheme = ({ children }: any) => {
    const theme = createTheme(RANKING_APP_THEME_MOCK);
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
        return (
            <MockTheme>
                <Provider store={store}>
                    <DndProvider
                        backend={TouchBackend}
                        options={{ enableMouseEvents: true }}
                    >
                        {children}
                    </DndProvider>
                </Provider>
            </MockTheme>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
