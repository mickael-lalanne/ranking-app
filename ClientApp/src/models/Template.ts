import { Element } from './Element';
import { UserId } from './User';

export interface Template {
    id?: string;
    name: string;
    userId?: UserId;
    tiers: Tier[];
    elements: Element[];
    createdAt?: string; // ex: "2023-11-16T20:26:59.654363Z"
}

export interface Tier {
    id?: string;
    name: string;
    rank: number;
}

export const TIERS_COLORS: string[] = [
    '#FF8C7F', // Rank 1
    '#FFC07D', // Rank 2
    '#FEF782', // Rank 3
    '#7AEA81', // Rank 4
    '#78BDFD', // Rank 5
];

export enum ETemplateMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

export enum EEditViewMode {
    EditElement = 'edit-element', // When user wants to edit an element
    EditTier = 'edit-tier', // When user wants to edit a tier
    Hide = 'hide' // When the view isn't displayed or an element
};
