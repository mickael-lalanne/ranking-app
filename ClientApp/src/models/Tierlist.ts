import { RankedElement } from './Element';
import { UserId } from './User';

export interface Tierlist {
    id: string;
    name: string;
    userId: UserId;
    templateId: string;
    rankedElements: RankedElement[];
    createdAt: string; // ex: "2023-11-16T20:26:59.654363Z"
}

export enum ETierlistDragItem {
    Element = 'element'
}
