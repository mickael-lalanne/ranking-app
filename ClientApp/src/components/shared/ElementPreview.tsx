import React, { useState, useEffect } from 'react';
import { css } from '@emotion/css';
import { Element } from '../../models/Element';
import RemoveCircleOutlineIcon from '@mui/icons-material/RemoveCircleOutline';
import { getElementImageUrl } from '../../services/CloudinaryService';
import { ELEMENT_SIZE } from '../../utils/css-utils';

const ElementPreview = ({
    element,
    deleteElementHandler, dragStartHandler, dragEndHandler,
    readonly = false
}: {
    element: Element,
    deleteElementHandler?: (elementId: number) => void,
    dragStartHandler?: (element: Element) => void,
    dragEndHandler?: () => void,
    readonly?: boolean
}) => {
    const [isHovering, setIsHovering] = useState<boolean>(false);
    const [elementImage, setElementImage] = useState<string>();

    // Called when the component is initialized and when the element prop has changed
    useEffect(() => {
        if (element.image) {
            const elementImage: string = !element.image.startsWith('data:image/png')
                // If file don't start with data:image/png, it means the image has already been uploaded to Cloudinary
                ? getElementImageUrl(element.image)
                // Otherwise, it means the element is not created yet, so use the data URI of the resized image
                : element.image;
            setElementImage(elementImage);
        }
    }, [element]);

    // Show the delete tier button only on hover
    const DeleteElementButton = (elementId: number): React.JSX.Element | undefined => {
        if (isHovering && deleteElementHandler) {
            return (
                <div className={element_delete_container_style}>
                    <RemoveCircleOutlineIcon
                        onClick={() => deleteElementHandler(elementId)}
                        className={remove_element_icon_style}
                    />
                    <div className={element_delete_bg_style}></div>
                </div>
            );
        }
    };

    return(<>
        <div
            className={element_container_style}
            onMouseEnter={() => setIsHovering(true)}
            onMouseLeave={() => setIsHovering(false)}
            onDragStart={e => dragStartHandler ? dragStartHandler(element) : undefined}
            onDragEnd={() => dragEndHandler ? dragEndHandler() : undefined}
            onDragOver={e => e.preventDefault()}
            draggable={!readonly}
        >
            <img src={elementImage} className={element_img_style} />
            {DeleteElementButton(element.id!)}
        </div>
    </>);
};

export default ElementPreview;

/**
 * CSS STYLES
 */
export const element_container_style = css({
    display: 'flex',
    alignItems: 'center',
    position: 'relative',
    border: '1px solid',
    flexDirection: 'column',
    height: ELEMENT_SIZE,
    justifyContent: 'center',
    textAlign: 'center',
    aspectRatio: 1,
    width: ELEMENT_SIZE
});

const element_delete_container_style = css({
    position: 'absolute',
    top: '0',
    left: '0',
    width: '100%',
    height: '100%',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    zIndex: 1
});

const element_delete_bg_style = css({
    position: 'absolute',
    top: '0',
    left: '0',
    width: '100%',
    height: '100%',
    opacity: 0.7,
    backgroundColor: 'black',
    zIndex: -1
});

const remove_element_icon_style = css({
    color: 'white',
    width: '65px !important',
    height: '65px !important',
    cursor: 'pointer'
});

const element_img_style = css({
    maxWidth: '100%',
    maxHeight: '100%',
    height: '100%',
    width: '100%',
    objectFit: 'cover'
});
