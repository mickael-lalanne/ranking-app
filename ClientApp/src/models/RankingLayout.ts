import { Template } from './Template';
import { Tierlist } from './Tierlist';
import TemplatesViewer from '../components/template/TemplatesViewer';
import TemplateEditor from '../components/template/TemplateEditor';
import { createTemplate, deleteTemplate, updateTemplate } from '../services/TemplateServices';
import TierlistsViewer from '../components/tierlist/TierlistsViewer';
import TierlistsEditor from '../components/tierlist/TierlistsEditor';
import { AppDispatch } from '../app/store';
import { createTierlist, deleteTierlist, updateTierlist } from '../services/TierlistServices';

export type RankingType = Template | Tierlist;

export enum ERankingLayoutMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

type EditorComponent = ({ saveHandler, itemToEdit, mode }: EditorComponentProps) => React.JSX.Element;

export interface EditorComponentProps {
    saveHandler: (itemToSave: Template | Tierlist) => Promise<void>;
    itemToEdit?: Template | Tierlist | undefined;
    mode: ERankingLayoutMode;
};

export interface RankingLayoutProps {
    viewerTitle: string;
    viewerSubtitle: string;
    viewerBtnText: string;
    builderTitle: string;
    builderSubtitle: string;
    builderBtnText: string;
    editorTitle: string;
    editorSubtitle: string;
    editorBtnText: string;
    ViewerComponent: ({ editHandler }: {
        editHandler: (template: Template) => void;
    }) => React.JSX.Element;
    EditorComponent: EditorComponent;
    createFunction: (itemToCreate: RankingType, dispatch: AppDispatch) => Promise<any>;
    updateFunction: (itemToUpdate: RankingType, dispatch: AppDispatch) => Promise<any>;
    deleteFunction: (itemIdToDelete: number, dispatch: AppDispatch) => Promise<any>;
}


export const TEMPLATE_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My templates',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add template',
    builderTitle: 'Template creation',
    builderSubtitle: 'Add a name, tiers and elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Template edition',
    editorSubtitle: 'Update your template here',
    editorBtnText: 'Cancel',
    ViewerComponent: TemplatesViewer,
    EditorComponent: TemplateEditor,
    createFunction: createTemplate as (templateToCreate: RankingType, dispatch: AppDispatch) => Promise<Template>,
    updateFunction: updateTemplate as (templateToUpdate: RankingType, dispatch: AppDispatch) => Promise<void>,
    deleteFunction: deleteTemplate
};

export const TIERLIST_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My tierlists',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add tierlist',
    builderTitle: 'Tierlist creation',
    builderSubtitle: 'Rank your elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Tierlist edition',
    editorSubtitle: 'Update your tierlist here',
    editorBtnText: 'Cancel',
    ViewerComponent: TierlistsViewer,
    EditorComponent: TierlistsEditor,
    createFunction: createTierlist as (tierlistToCreate: RankingType) => Promise<Tierlist>,
    updateFunction: updateTierlist as (tierlistToUpdate: RankingType) => Promise<Response>,
    deleteFunction: deleteTierlist
};
