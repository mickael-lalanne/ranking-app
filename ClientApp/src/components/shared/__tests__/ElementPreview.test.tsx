import * as React from 'react';
import { renderWithProviders } from '../../../utils/test-utils';
import { fireEvent, screen } from '@testing-library/react';
import ElementPreview, { ElementPreviewProps } from '../ElementPreview';
import { Element } from '../../../models/Element';
import userEvent from '@testing-library/user-event';

const elementSample: Element = {
    id: 'element id',
    name: 'element name',
    image: 'element image',
};

const DEFAULT_PROPS: ElementPreviewProps = {
    element: elementSample,
};

const ElementPreviewTest = (props: ElementPreviewProps) => {
    return (
        <ElementPreview
            element={props.element}
            deleteElementHandler={props.deleteElementHandler}
            readonly={props.readonly}
            fitToContainer={props.fitToContainer}
        />
    );
};

describe('Screenshot', () => {
    it('takes default screenshot', () => {
        const component = renderWithProviders(
            ElementPreviewTest(DEFAULT_PROPS)
        );

        expect(component.baseElement).toMatchSnapshot();
    });

    it('takes screenshot with delete button', async () => {
        const component = renderWithProviders(
            ElementPreviewTest({
                ...DEFAULT_PROPS,
                deleteElementHandler: jest.fn(),
            })
        );

        // Hover the element preview
        await userEvent.hover(screen.getByTestId('element-preview'));

        expect(component.baseElement).toMatchSnapshot('With delete button');
    });
});

describe('Common', () => {
    it('deletes element', async () => {
        const deleteMock: jest.Mock = jest.fn();
        renderWithProviders(
            ElementPreviewTest({
                element: elementSample,
                deleteElementHandler: deleteMock,
            })
        );

        // The delete button should not be displayed
        expect(screen.queryByTestId('delete-element')).toBe(null);

        // Hover the element preview
        await userEvent.hover(screen.getByTestId('element-preview'));

        // With the delete button now displayed, click on it
        await userEvent.click(screen.getByTestId('RemoveCircleOutlineIcon'));
        expect(deleteMock).toHaveBeenCalledWith(elementSample.id);
    });
});

