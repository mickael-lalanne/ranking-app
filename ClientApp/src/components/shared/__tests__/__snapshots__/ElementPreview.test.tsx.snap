// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes default screenshot 1`] = `
<body>
  <div>
    <div
      class="css-13ambje"
      data-cy="element-preview"
      data-testid="element-preview"
      draggable="true"
      style="height: 75px; width: 75px; opacity: 1;"
    >
      <img
        class="css-40zi5q"
        data-cy="element-preview-img"
        src="https://res.cloudinary.com/ranking-app/image/upload/element image"
      />
    </div>
  </div>
</body>
`;

exports[`Screenshot takes screenshot with delete button: With delete button 1`] = `
<body>
  <div>
    <div
      class="css-13ambje"
      data-cy="element-preview"
      data-testid="element-preview"
      draggable="true"
      style="height: 75px; width: 75px; opacity: 1;"
    >
      <img
        class="css-40zi5q"
        data-cy="element-preview-img"
        src="https://res.cloudinary.com/ranking-app/image/upload/element image"
      />
      <div
        class="css-esyjz6"
        data-testid="delete-element"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-cxyv8p css-i4bv87-MuiSvgIcon-root"
          data-testid="RemoveCircleOutlineIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 11v2h10v-2H7zm5-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
        <div
          class="css-10le4f0"
        />
      </div>
    </div>
  </div>
</body>
`;
